### DISCUSSION TITLE
ts problem with writeAsync returned function

### DISCUSSION THREAD

**Discussion by variousred** [wevm/wagmi#1949] [Q&A] [answered]
destructuring the response from useContractWrite and assigning the `writeAsync` function, how can you later call that function with arguments and not have TS complain that it expected 0 arguments?

`  const {
    writeAsync,
    write,
    ...status
  } = useContractWrite(config)`
  and then
  `    await writeAsync?.({
      recklesslySetUnpreparedArgs: [
        aaa,
        bbb,
        ccc,
        ddd,
        false,
        customMessage.replace(/[^a-zA-Z0-9 .?!/:]/g, '')
      ]
    })`
    I get 
    
![image](https://user-images.githubusercontent.com/238354/222918785-974a73b4-da36-45cf-bbca-7f8258daa105.png)

ts 4.9.5

**Comment by tmm** [comment, marked as answer]
Looks like you are using `usePrepareContractWrite`. When you prepare a transaction ahead of time (recommended), you can't pass in dynamic `args` to `useContractWrite`'s `writeAsync` function because they are not part of the prepared transaction. Instead of passing in dynamic `args` to `writeAsync`, you likely want to update the `args` that are passed to `usePrepareContractWrite` with a reactive state variable instead.

Below is an example of the differences between prepared and unprepared contract writes:

```ts
import { useContractWrite, usePrepareContractWrite, erc20ABI } from 'wagmi'
import { parseAbi } from 'abitype'

const abi = parseAbi(['function foo(address, string)'])

const { config } = usePrepareContractWrite({
  abi,
  functionName: 'foo',
  args: ['0x123', 'test'],
})

const { writeAsync: preparedWriteAsync } = useContractWrite(config)
//                  ^? const preparedWriteAsync: (() => Promise<SendTransactionResult>) | undefined
preparedWriteAsync?.()

const { writeAsync: unpreparedWriteAsync } = useContractWrite({
  //                ^? const unpreparedWriteAsync: (config: { recklesslySetUnpreparedArgs?: readonly [`0x${string}`, string] | undefined; recklesslySetUnpreparedOverrides?: (Overrides & { from?: `0x${string}` | undefined; }) | undefined }) => Promise<SendTransactionResult>
  mode: 'recklesslyUnprepared',
  abi,
  functionName: 'foo',
  args: ['0x123', 'test'],
})
unpreparedWriteAsync({
  recklesslySetUnpreparedArgs: ['0x123', 'test'],
})
```

[TypeScript Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgVwM4FMDCEB2MoCGAxjAOpTAzoA0KGAClOmAU9noSeZTXOlEQBMABgCCAIQCScAL5wAZlAgg4AcgDuBAOYhgqgFChIsRHBZQMogEbBZCpStUEbMAJ5h0B-UVyp4z2wBeM1ZLGwAKAG1VeWQcEmBcBQgIcIIAE3SmVFRaPwocLQBKVQBdIv1vX3gkHxx5YC07YLR0RmZWLFx8YjIKKnCEfTg4AOphhTiE3AA5AhB0AC41eRTVcZHWLVRl6OEADwBGQQBmdbUqPzLxmQqqnD9TdX70UVRXeOWwJnN0dO4qG8PkRmnQuhxegD0OE6g1ivoAPQIkYo1Fo1EAPQA-Ppvh0mP8XkD4liAHThO51R5IZ48YlEZZxPG-Ql097xUGtdg9LgvQYTJHooUo7ETEAQdJLNRMIgAawANugcvLXABVHDMzrpdYTMYTWLxGCJHBzBbLGJrDajKDbXaqA7HM60VSXGDXfS3fRMn5aqH0-kjGUKpWoFUAZXQMHVmoJohtOzgeyOp3OLqVbtKNwqQA)
